File: src/services/purchase-invoice.service.ts
import api from "./api";
import {
  PurchaseInvoice,
  CreatePurchaseInvoiceRequest,
  ReceiveOrderRequest,
  RejectOrderRequest,
  ReceiveSummaryResponse,
  ReceiveOrderFilters,
} from "@/types/purchase-invoice.types";
import {
  ApiResponse,
  PaginatedResponse,
} from "@/types/api.types";

function buildParams(filters?: ReceiveOrderFilters) {
  if (!filters) return undefined;
  const out: Record<string, string | number> = {};
  if (filters.page !== undefined) out.page = filters.page;
  if (filters.size !== undefined) out.size = filters.size;
  if (filters.search) out.search = filters.search;
  if (filters.receiptStatus) out.receiptStatus = filters.receiptStatus;
  if (filters.supplierId !== "" && filters.supplierId !== undefined) out.supplierId = filters.supplierId;
  if (filters.warehouseId !== "" && filters.warehouseId !== undefined) out.warehouseId = filters.warehouseId;
  if (filters.dateFrom) out.dateFrom = filters.dateFrom;
  if (filters.dateTo) out.dateTo = filters.dateTo;
  return out;
}

export const purchaseInvoiceService = {
  getAll: (params?: ReceiveOrderFilters) =>
    api.get<PaginatedResponse<PurchaseInvoice>>("/purchase-invoices", {
      params: buildParams(params),
    }),

  getById: (id: number) =>
    api.get<ApiResponse<PurchaseInvoice>>(`/purchase-invoices/${id}`),

  create: (data: CreatePurchaseInvoiceRequest) =>
    api.post<ApiResponse<PurchaseInvoice>>("/purchase-invoices", data),

  delete: (id: number) =>
    api.delete<ApiResponse<void>>(`/purchase-invoices/${id}`),

  approve: (id: number) =>
    api.post<ApiResponse<PurchaseInvoice>>(`/purchase-invoices/${id}/approve`, {}),

  receive: (id: number, data: ReceiveOrderRequest) =>
    api.post<ApiResponse<PurchaseInvoice>>(`/purchase-invoices/${id}/receive`, data),

  reject: (id: number, data: RejectOrderRequest) =>
    api.post<ApiResponse<PurchaseInvoice>>(`/purchase-invoices/${id}/reject`, data),

  getReceiveSummary: () =>
    api.get<ApiResponse<ReceiveSummaryResponse>>("/purchase-invoices/receive-summary"),
};

File: inflow-backend/src/main/java/com/john/inflow/service/PurchaseInvoiceService.java
package com.john.inflow.service;

import com.john.inflow.dto.request.CreatePurchaseInvoiceRequest;
import com.john.inflow.dto.request.ReceiveOrderRequest;
import com.john.inflow.dto.request.RejectOrderRequest;
import com.john.inflow.dto.response.PageResponse;
import com.john.inflow.dto.response.PurchaseInvoiceResponse;
import com.john.inflow.dto.response.ReceiveSummaryResponse;

import java.time.OffsetDateTime;

public interface PurchaseInvoiceService {
    PurchaseInvoiceResponse create(CreatePurchaseInvoiceRequest request, Integer userId);
    PurchaseInvoiceResponse getById(Integer id);
    PageResponse<PurchaseInvoiceResponse> getAll(int page, int size);
    PageResponse<PurchaseInvoiceResponse> search(
            String search,
            String receiptStatus,
            Integer supplierId,
            Integer warehouseId,
            OffsetDateTime dateFrom,
            OffsetDateTime dateTo,
            int page,
            int size
    );
    void delete(Integer id);

    PurchaseInvoiceResponse receive(Integer id, ReceiveOrderRequest request, Integer userId);
    PurchaseInvoiceResponse reject(Integer id, RejectOrderRequest request, Integer userId);
    ReceiveSummaryResponse getReceiveSummary();
}
